### PR TITLE
[Stashing] Stash Diff Viewer

### PR DESCRIPTION
**Pull request by outofambit**
🌵 🌵 **Review this _before_ https://github.com/desktop/desktop/pull/7210** 🌵 🌵 

(the primary intent here is to get @iAmWillShepherd's thoughts on this before #7210 is reviewed/merged, but others' feedback is certainly welcome and helpful!)

## Overview

This PR adds the ability to view the contents of a stash (files and their diffs) to the new stashing workflow.

gif

### Implementation

* adds Stash Diff Viewer UI component (based heavily on history commit viewer components)
* adds app UI state for the Stash Diff Viewer component
* adds dispatcher/app-store/git-store glue between components and git functionality
* adds state to stashEntry to include a list of files touched by that stash

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -4751,6 +4759,35 @@ export class AppStore extends TypedBaseStore<IAppState> {
     log.info(`Popped stash with commit id ${stash.stashSha}`)
   }
 
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public async _popStashEntry(repository: Repository, stashEntry: IStashEntry) {
+    const gitStore = this.gitStoreCache.get(repository)
+    await gitStore.performFailableOperation(() => {
+      return popStashEntry(repository, stashEntry.stashSha)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/app-store.ts]: Good call wrapping this in `GitStore.performFailableOperation`.

File: app/src/models/stash-entry.ts [hunk 2]
```diff
@@ -7,4 +9,16 @@ export interface IStashEntry {
 
   /** The SHA of the commit object created as a result of stashing. */
   readonly stashSha: string
+
+  /** The list of files this stash touches */
+  readonly files: StashedFileChanges
 }
+
+export enum StashedChangesLoadStates {
```
Comment (iAmWillShepherd) [inline app/src/models/stash-entry.ts]: This can use some 📝 since it's exported.

File: app/src/models/stash-entry.ts [hunk 3]
```diff
@@ -7,4 +9,16 @@ export interface IStashEntry {
...
+  NotLoaded,
+  Loading,
+}
+
+export type StashedFileChanges =
```
Comment (iAmWillShepherd) [inline app/src/models/stash-entry.ts]: 📝 since it's exported.

Comment (iAmWillShepherd) [inline app/src/models/stash-entry.ts]: This type feels like it's trying to accomplish two distinct things, how do you feel about making this an interface?

Ex.
```typescript
inteface ICouldNotThinkOfAGoodName {
  readonly files: ReadonlyArray<CommittedFileChanges>
  readonly loadState: StashedChangesLoadState
}
```

File: app/src/ui/repository.tsx [hunk 4]
```diff
@@ -283,15 +285,31 @@ export class RepositoryView extends React.Component<
...
-        return (
-          <div>
-            <p>
-              {`${stashEntry.name} created on ${
-                stashEntry.branchName
-              }@${stashEntry.stashSha.substr(0, 7)}`}
-            </p>
-          </div>
-        )
+        if (Array.isArray(stashEntry.files)) {
```
Comment (iAmWillShepherd) [inline app/src/ui/repository.tsx]: Changing the type definition for `StashedFileChanges` will remove the need for this check.

File: app/src/lib/stores/app-store.ts [hunk 5]
```diff
@@ -4736,19 +4752,33 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+  ) {
+    const gitStore = this.gitStoreCache.get(repository)
+    await gitStore.performFailableOperation(() => {
+      return dropDesktopStashEntry(repository, stashEntry.stashSha)
+    })
+    log.info(`Dropped stash with commit id ${stashEntry.stashSha}`)
+
+    await this._refreshRepository(repository)
+
+    this.emitUpdate()
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: Does this `emitUpdate` need to happen? I believe `_refreshRepository()` should cover this...

File: app/src/models/stash-entry.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/models/stash-entry.ts]: ah yes i'm into it! will update now

**Comment by shiftkey** [comment]
Does this need a separate reviewer or are you on top of this @iAmWillShepherd?

**Comment by iAmWillShepherd** [comment]
> Does this need a separate reviewer or are you on top of this @iAmWillShepherd?

I got it. Thanks for popping in and adding some additional review points. ⚡️

File: app/src/models/stash-entry.ts (hunk 3 above)
Comment (iAmWillShepherd) [reply app/src/models/stash-entry.ts]: Go with https://github.com/desktop/desktop/pull/7213#discussion_r274547593

File: app/src/lib/stores/app-store.ts [hunk 6]
```diff
@@ -4736,19 +4752,39 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+      }`
     )
 
-    if (stash === null) {
-      log.warn('no stash found that matches')
-      return
-    }
+    await this._refreshRepository(repository)
+
+    this.emitUpdate()
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: I think this one can also be dropped:

```suggestion
```

File: app/src/lib/stores/app-store.ts (hunk 5 above)
Comment (iAmWillShepherd) [reply app/src/lib/stores/app-store.ts]: Looks like we do need it.

https://github.com/desktop/desktop/blob/974b0a4790d46e82a4f2e99ffe2d7bdef8efd596/app/src/lib/stores/app-store.ts#L2206-L2259

My question is should we be emitting an update when we refresh the repo?

File: app/src/models/stash-entry.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/models/stash-entry.ts]: done in 2aaf6feb5ec981f0a98eb8a5497dda62e949d375. thank you!

File: app/src/lib/stores/app-store.ts (hunk 6 above)
Comment (outofambit) [reply app/src/lib/stores/app-store.ts]: removed in https://github.com/desktop/desktop/pull/7213/commits/3c9e4335dec0ae98f170b867f562450da439f783

**Pull request merged**